File: src/test/scala/com/scala/scogen/datagenerator/EmployeeFinanceDatasetTest.scala
package com.scala.scogen.datagenerator

import com.scala.scogen.BaseClassTest
import com.scala.scogen.metaData.EmployeeNameList
import org.apache.spark.sql.{Column, Dataset, Row}
import org.junit.Assert

class EmployeeFinanceDatasetTest extends BaseClassTest {

  var employeeFinanceDataset: Dataset[Row] = spark.emptyDataFrame
  import spark.implicits._

  def initilize(): Unit =
  {
    if(employeeFinanceDataset.isEmpty) {
      val listOfEmployees = EmployeeNameList.setofUniqueNames
      val employeeDs = new EmployeeDetailsDataset().employeeTable(spark, listOfEmployees)

      val eids = employeeDs.rdd.map(x => x.getAs[String]("empid")).collect().toList
      employeeFinanceDataset = new EmployeeFinanceDataset().employeeFinanceData(spark, eids).cache()
    }
  }

  "finance data" should "also have only 1000 employes" in {
    initilize()
    Assert.assertEquals(1000, employeeFinanceDataset.count)
  }

  "employee ctc range" should "be 10,000 - 1,00,000" in {
    initilize()
    val count = employeeFinanceDataset
      .filter($"ctc" > 100000 && $"basic"<10000).count
    Assert.assertEquals(0, count)

  }

  "all eployees " should "have basic as 20% of ctc not more" in {

    initilize()
    val result = employeeFinanceDataset
      .select($"ctc",$"basic").rdd.map((x: Row) => {
      def foo(x: Row) = {
        val ctc = x.getAs[Integer]("ctc")
        val basic = x.getAs[Double]("basic")
        basic == ctc * 0.2
      }

      foo(x)
    }).reduce((x: Boolean, y: Boolean) => x && y)

    Assert.assertTrue(result)
  }



}

File: src/test/java/com/java/scogen/datagenerator/EmployeeFinanceDatasetTest.java
package com.java.scogen.datagenerator;

import com.java.scogen.SparkTestSessionGenerator;
import com.java.scogen.metadata.EmployeeNameList;
import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;
import java.util.Set;

public class EmployeeFinanceDatasetTest extends SparkTestSessionGenerator {


    private Dataset<Row> employeeFinanceDataset = null;

    private void initilize()
    {
        if(employeeFinanceDataset == null) {
            EmployeeNameList employeeNameList = new EmployeeNameList();
            Set<String> nameSet = employeeNameList.setofUniqueNames();
            EmployeeDetailsDataset employeeDetailsDataset = new EmployeeDetailsDataset(getSpark());
            Dataset<Row> employee = employeeDetailsDataset.EmployeeTable(nameSet);
            List<String> ids = employee.select("empid").toJavaRDD().map(x -> x.<String>getAs("empid"))
                    .collect();
            employeeFinanceDataset = new EmployeeFinanceDataset(getSpark())
                    .employeeFinanceData(ids).cache();
        }

    }

    @Test
    public void employeeCountTest()
    {
        initilize();
        Assert.assertEquals(1000, employeeFinanceDataset.count());

    }

    @Test
    public void employeeCTCRangeTest()
    {
        initilize();
        long count = employeeFinanceDataset.filter(new Column("ctc").$greater(1_00_000)
                .and(new Column("basic").$less(10_000))).count();
        Assert.assertEquals(0, count);

    }

    @Test
    public void employeesBasicRangeTest()
    {
        initilize();
        boolean result = employeeFinanceDataset.select(new Column("ctc"),new Column("basic"))
                .toJavaRDD()
                .map(x-> {
                    int ctc = x.<Integer>getAs("ctc");
                    double basic = x.<Double>getAs("basic");
                    return basic == ctc*0.2;
                }).reduce((x,y)-> x && y);

        Assert.assertTrue(result);

    }

}
